### DISCUSSION TITLE
Bitbucket Unit Testing

### DISCUSSION THREAD

**Discussion by DeniseSkidmore** [pythonnet/pythonnet#2194] [Q&A] [answered]
I can write C# and python yml steps in bitbucket, but combining the two correctly is eluding me.  How should this be set up in BitBucket Yml?

Current partial solution:
`image: python:3.11`

(step with image: mcr.microsoft.com/dotnet/sdk:6.0 sets up the dll artifacts, which I can access below)
```
          script:
            - if [ -f unit_test_requirements.txt ]; then pip install -r unit_test_requirements.txt; fi
            # https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#scripted-install
            - wget https://dot.net/v1/dotnet-install.sh -O dotnet-install.sh
            - chmod +x ./dotnet-install.sh
            - ./dotnet-install.sh --runtime dotnet --version latest
            - export DOTNET_ROOT=$HOME/.dotnet
            - export PATH=$PATH:$DOTNET_ROOT:$DOTNET_ROOT/tools
            - export PYTHONPATH=$PYTHONPATH:$DOTNET_ROOT:$DOTNET_ROOT/tools         
            - cd PythonLib/Tests/UnitTests
            - python -m unittest *_test.py
```

this dotnet install gets me past the loadclr and AddReference steps, but the first time a function is called that relies on a system dll, I get:
`System.IO.FileLoadException: Could not load file or assembly 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. Could not find or load a specific file. (0x80131621)`

At my coworker's recommendation, I inverted the image and install steps.

```
          image: mcr.microsoft.com/dotnet/sdk:6.0
          caches:
            - dotnetcore
          script:
            - apt-get update
            - apt-get install -y binutils
            - apt-get install -y python3
            - apt-get install -y python3-pip
```
I end up getting the exact same error after all that.
```
System.IO.FileLoadException: Could not load file or assembly 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at Python.Runtime.AssemblyManager.ResolveHandler(Object ob, ResolveEventArgs args) in /tmp/build-via-sdist-dq4gmg3h/pythonnet-3.0.1/src/runtime/AssemblyManager.cs:line 113
   at System.Runtime.Loader.AssemblyLoadContext.InvokeResolveEvent(ResolveEventHandler eventHandler, RuntimeAssembly assembly, String name)
   at System.Runtime.Loader.AssemblyLoadContext.OnAssemblyResolve(RuntimeAssembly assembly, String assemblyFullName)
```

**Comment by lostmsu** [comment]
What does the error look like?

**Reply by DeniseSkidmore** [reply]
```
System.IO.FileLoadException: Could not load file or assembly 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.

   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)

   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)

   at System.Reflection.Assembly.LoadFrom(String assemblyFile)

   at Python.Runtime.AssemblyManager.ResolveHandler(Object ob, ResolveEventArgs args) in /tmp/build-via-sdist-dq4gmg3h/pythonnet-3.0.1/src/runtime/AssemblyManager.cs:line 113

   at System.Runtime.Loader.AssemblyLoadContext.InvokeResolveEvent(ResolveEventHandler eventHandler, RuntimeAssembly assembly, String name)

   at System.Runtime.Loader.AssemblyLoadContext.OnAssemblyResolve(RuntimeAssembly assembly, String assemblyFullName)
```

**Reply by lostmsu** [reply]
Sounds like your script does not configure .NET runtime. See https://pythonnet.github.io/pythonnet/python.html#loading-a-runtime

Alternatively if you are just trying to use .NET system libraries from Python (e.g. not your own), your code fails because these system libraries are not part of the .NET distribution and need to be shipped along with your Python script.

**Reply by DeniseSkidmore** [reply]
I am calling load('coreclr') in my script.  Is it maybe the wrong runtime for this environment?  I didn't even get past 
load('coreclr') until I added the dotnet install lines above, so I think it's at least partially loading the runtime.  I think it's more about how the runtime is installed?  (which is your second point, but still isn't a solution)  The System.Text.Json dll is in my folder of nuget packages downloaded even if it's not in the runtime itself, you would think it should be able to find it?

**Reply by DeniseSkidmore** [reply]
Which brings up the important detail in that I'm getting my custom dlls thusly, but should be migrating towards some sort of wheel/pip install?
`sys.path.append(r'../../../CSharpLib/bin/Debug/net6.0')`

**Reply by DeniseSkidmore** [reply]
Mono load gets less far, pretty sure the dotnet is the right load for that install, I'm just missing an environment variable or install step...
```
Traceback (most recent call last):

  File "/usr/local/lib/python3.11/site-packages/pythonnet/__init__.py", line 75, in _create_runtime_from_spec

    return clr_loader.get_mono(**params)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.11/site-packages/clr_loader/__init__.py", line 73, in get_mono

    libmono = find_libmono(sgen=sgen)

              ^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.11/site-packages/clr_loader/util/find.py", line 145, in find_libmono

    raise RuntimeError("Could not find libmono")

RuntimeError: Could not find libmono
```

**Reply by DeniseSkidmore** [reply]
So I tried comparing runtimes between Bitbucket server and my PC
Bitbucket
```
+ python -c "import pythonnet; pythonnet.load('coreclr'); print(pythonnet.get_runtime_info())"
Runtime: CoreCLR
=============
  Version:      <undefined>
  Initialized:  True
  Shut down:    False
  Properties:
    System.Reflection.Metadata.MetadataUpdater.IsSupported = false
    RUNTIME_IDENTIFIER = debian.12-x64
    FX_DEPS_FILE = /root/.dotnet/shared/Microsoft.NETCore.App/6.0.20/Mi…
    APP_CONTEXT_DEPS_FILES = /root/.dotnet/shared/Microsoft.NETCore.App…
    APP_CONTEXT_BASE_DIRECTORY = 
    PLATFORM_RESOURCE_ROOTS = :
    PROBING_DIRECTORIES = 
    NATIVE_DLL_SEARCH_DIRECTORIES = :/root/.dotnet/shared/Microsoft.NET…
    TRUSTED_PLATFORM_ASSEMBLIES = /root/.dotnet/shared/Microsoft.NETCor…
```

My PC:
```
C:\Users\denise_skidmore>python -c "import pythonnet; pythonnet.load('coreclr'); print(pythonnet.get_runtime_info())"
Runtime: CoreCLR
=============
  Version:      <undefined>
  Initialized:  True
  Shut down:    False
  Properties:
    FX_DEPS_FILE = C:\Program Files\dotnet\shared\Microsoft.NETCore.App…
    TRUSTED_PLATFORM_ASSEMBLIES = C:\Program Files\dotnet\shared\Micros…
    NATIVE_DLL_SEARCH_DIRECTORIES = ;C:\Program Files\dotnet\shared\Mic…
    PLATFORM_RESOURCE_ROOTS = ;
    APP_CONTEXT_BASE_DIRECTORY =
    APP_CONTEXT_DEPS_FILES = C:\Program Files\dotnet\shared\Microsoft.N…
    PROBING_DIRECTORIES =
    RUNTIME_IDENTIFIER = win10-x64
    System.Reflection.Metadata.MetadataUpdater.IsSupported = false

```
It's expected that they are different,

**Reply by DeniseSkidmore** [reply]
A [similar issue](https://github.com/pythonnet/pythonnet/issues/1377) 

- Suggests explicit AddReference of the dependency
    - I got the same error `System.IO.FileLoadException: Could not load file or assembly 'System.Text.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.` but on the add reference line instead of on first usage.  
    - The fact that it found the version number and the public key token implies that it's a dependency not System.Text.Json itself failing to load.
    - Nuget only lists System.Text.Encodings.Web as a dependency of System.Text.Json for .NET 7.
    - explicitly loading System.Text.Encodings.Web has the exact same issue `System.IO.FileLoadException: Could not load file or assembly '[System.Text.Encodings.Web](http://system.text.encodings.web/), Version=7.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.`
- points to a [root cause](https://github.com/python/cpython/issues/83086) in python itself.
    - Trying earlier python version did not solve the issue, so perhaps my issue is unrelated.

**Reply by DeniseSkidmore** [reply]
Earlier version of python didn't solve my issue.  Inverting my setup to dotnet image and manual python install didn't solve my issue.

**Comment by DeniseSkidmore** [comment, marked as answer]
@lostmsu was on the right track although some context was missing.

.NET applications typically use a runtimeconfig.json file to set the runtime version the application is compiled for.  In normal C# development this is automatically generated and linked by the compiler and you never need look at it.  I compiled my nuget packages for .NET Standard 2.0 and .NET 6, and had the .NET 6 runtime installed, which all sounds compatible, except some of my dependencies explicitly work with either .NET 6 or .NET 7, and with no runtimeconfig.json to determine which to run, it was trying .NET 7 and failing to find the runtime.  It worked on my local machine because I had the .NET 7 runtime installed.  

[The wiki that @lostmsu linked](https://pythonnet.github.io/pythonnet/python.html#loading-a-runtime) mentions runtimeconfig, and how to explicitly set one, but I didn't think I needed that since the code itself was running fine in some environments, I didn't think it was the code.  I was loading the runtime as the the [readme](https://github.com/pythonnet/pythonnet/#net-core) recommends.  I have not tested the explicit runtimeconfig solution, I just installed the .NET 7 runtime, as our environment is moving in that direction anyway.

Secondarily, and important if anyone reads this because of the BitBucket keyword, the install process I used initially doesn't grab the latest .NET, and may leave out some dependencies.  The `--version latest` flag gets version 6, not 7.  You can get the actual OS version running, then use the [distribution specific package install](https://learn.microsoft.com/en-us/dotnet/core/install/linux-debian#debian-12), (with minor modification) to install .NET 7 runtime.  I also tested this new install method with explicitly loading .NET 6 runtime and brought back the failure, so I was loading the .NET 7 runtime in my code correctly already, and it wasn't just the potential dependencies issue.

            # You need to know the specific linux version - cat /etc/os-release
            # https://learn.microsoft.com/en-us/dotnet/core/install/linux-debian#debian-12
            - wget https://packages.microsoft.com/config/debian/12/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
            - dpkg -i packages-microsoft-prod.deb
            - rm packages-microsoft-prod.deb
            - apt-get update && apt-get install -y dotnet-runtime-7.0
            # confirm install location to use in DOTNET_ROOT - dotnet --list-runtimes
            - export DOTNET_ROOT=/usr/share/dotnet/
            - export PATH=$PATH:$DOTNET_ROOT:$DOTNET_ROOT/tools
            - export PYTHONPATH=$PYTHONPATH:$DOTNET_ROOT:$DOTNET_ROOT/tools:$BITBUCKET_CLONE_DIR/MyLibArtifactLocation

**Reply by DeniseSkidmore** [reply]
[More info on runtimeconfig](https://learn.microsoft.com/en-us/dotnet/core/runtime-config/#runtimeconfigjson) - but you can probably just copy your config from the C# project where you built or downloaded the dlls.

**Reply by lostmsu** [reply]
In short the best way to get a suitable runtime config is to build a .NET exe with all the dependencies you want to load with `--publish` and take the resulting config.
